src/CharterManager.sol

```diff
@@ -0,0 +1,253 @@
+    bytes32 slot1;
+    mapping (address => address) public proxy; // UrnProxy per user
+    mapping (address => mapping (address => uint256)) public can;
+
+    mapping (bytes32 => uint256)                     public gate; // allow only permissioned vaults
+    mapping (bytes32 => uint256)                     public Nib;  // fee percentage for un-permissioned vaults [wad]
+    mapping (bytes32 => mapping(address => uint256)) public nib;  // fee percentage for permissioned vaults    [wad]
+    mapping (bytes32 => mapping(address => uint256)) public line; // debt ceiling for permissioned vaults      [rad]
+
+    address public immutable vat;
+    address public immutable vow;
+
+    // --- Administration ---
+    event File(bytes32 indexed ilk, bytes32 indexed what, bool data);
+    event File(bytes32 indexed ilk, bytes32 indexed what, uint256 data);
+    event File(bytes32 indexed ilk, address indexed user, bytes32 indexed what, uint256 data);
+    function file(bytes32 ilk, bytes32 what, uint256 data) external auth {
+        if (what == "gate") gate[ilk] = data;
+        else if (what == "Nib") Nib[ilk] = data;
+        else revert("CharterManager/file-unrecognized-param");
+        emit File(ilk, what, data);
+    }
+    function file(bytes32 ilk, address user, bytes32 what, uint256 data) external auth {
+        if (what == "line") line[ilk][user] = data;
+        else if (what == "nib") nib[ilk][user] = data;
+        else revert("CharterManager/file-unrecognized-param");
+        emit File(ilk, user, what, data);
+    }
+
+    // --- Math ---
+    uint256 constant WAD = 10 ** 18;
+
+    function sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "ds-math-sub-underflow");
+    }
+    function mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x);
+    }
+    function wmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = mul(x, y) / WAD;
+    }
+
+    // --- Auth ---
+    modifier auth { require(wards[msg.sender] == 1, "CharterManager/non-authed"); _; }
+
+    constructor(address vat_, address vow_) public {
+        vat = vat_;
+        vow = vow_;
+    }
+
+    event Allow(address indexed from, address indexed to);
+    event Disallow(address indexed from, address indexed to);
+    modifier allowed(address usr) {
+        require(msg.sender == usr || can[usr][msg.sender] == 1, "CharterManager/not-allowed");
+        _;
+    }
+    function allow(address usr) external {
+        can[msg.sender][usr] = 1;
+        emit Allow(msg.sender, usr);
+    }
+    function disallow(address usr) external {
+        can[msg.sender][usr] = 0;
+        emit Disallow(msg.sender, usr);
+    }
+
+    function getOrCreateProxy(address usr) public returns (address urp) {
+        urp = proxy[usr];
+        if (urp == address(0)) {
+            urp = proxy[usr] = address(new UrnProxy(address(vat), usr));
```

Comment:
I think we should probably emit an event whenever we create a new proxy.

Reply:
Done.